research_daps/orms/glass.py

```diff
@@ -0,0 +1,202 @@
+    # TODO: Active? boolean column indicating whether sector is in latest data dump
+    # TODO: Rank Integer column indicating sector rank
+    # TODO: Date Date column indicating date of latest dump with this relationship
+)
+
+notice_terms = Table(
+    "notice_terms",
+    Base.metadata,
+    Column("notice_id", Integer, ForeignKey("notice.notice_id")),
+    Column("term_id", Integer, ForeignKey("covid_term.term_id")),
+)
+
+
+class Organisation(Base):
+    """An organisation relating to a business website"""
+
+    org_id = Column(Integer, primary_key=True)
+    name = Column(
+        String,
+        unique=True,
+        doc="Organisation name inferred by named entity recognition",
+        index=True,
+    )
+    website = Column(String, nullable=False, doc="URL of organisation")
+    active = Column(
+        Boolean,
+        nullable=False,
+        doc="True if Organisation was contained in last data-dump",
+    )
+    # One-to-many
+    notices = relationship("Notice", backref=backref("organisation"))
+    descriptions = relationship(
+        "OrganisationDescription", backref=backref("organisation")
+    )
+    metadata_ = relationship("OrganisationMetadata", backref=backref("organisation"))
+    # Many-to-many
+    sectors = relationship(
+        "Sector", secondary=organisation_sector, back_populates="organisations"
+    )
+    addresses = relationship("OrganisationAddress", back_populates="organisation")
+    # addresses = relationship(
+    #     "Address", secondary=organisation_address, back_populates="organisations"
+    # )
+
+
+class OrganisationMetadata(Base):
+    """Organisation metadata which may not be stable over time"""
+
+    org_id = Column(Integer, ForeignKey("organisation.org_id"), primary_key=True)
+    date = Column(
+        Date, primary_key=True, doc="Date of data-dump inserting row", index=True
+    )
+    has_webshop = Column(
+        Boolean, nullable=False, doc="If True, presence of a webshop was found"
+    )
+    vat_number = Column(String, doc="VAT number (low population)")
+    low_quality = Column(
+        Boolean,
+        nullable=False,
+        doc="If True, information for `org_id` was of low quality",
+    )
+
+
+class OrganisationDescription(Base):
+    """Descriptions of organisation activities"""
+
+    # description_id = Column(Integer, primary_key=True)
+    org_id = Column(Integer, ForeignKey("organisation.org_id"), primary_key=True)
+    description = Column(
+        String, nullable=False, doc="Description of organisation extracted from website"
+    )
+    date = Column(
+        Date,
+        nullable=False,
+        primary_key=True,
+        index=True,
+        doc="Date of data-dump inserting row",
+    )
+
+
+class Address(Base):
+    """List of addresses found in websites"""
+
+    address_id = Column(Integer, primary_key=True)
+    address_text = Column(String, nullable=False, unique=True, doc="Full address text")
```

Comment:
The `unique` check on this is quite heavy, and unlikely to catch many exceptions (and even if they do, it probably won't change much?)

Reply:
or is this to trigger the autopk? either way, on ingestion it could get quite heavy then needing to generate the link table back to orgs